File: app/src/main/java/maderski/bluetoothautoplaymusic/ui/fragments/OptionsFragment.java
package maderski.bluetoothautoplaymusic.ui.fragments;

import android.Manifest;
import android.content.Context;
import android.graphics.Typeface;
import android.media.AudioManager;
import android.os.Build;
import android.os.Bundle;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.SeekBar;
import android.widget.Switch;
import android.widget.TextView;

import maderski.bluetoothautoplaymusic.analytics.FirebaseHelper;
import maderski.bluetoothautoplaymusic.helpers.PermissionHelper;
import maderski.bluetoothautoplaymusic.R;
import maderski.bluetoothautoplaymusic.sharedprefs.BAPMPreferences;

public class OptionsFragment extends Fragment {
    private static final String TAG = "OptionsFragment";

    private FirebaseHelper mFirebaseHelper;

    public OptionsFragment() {
        // Required empty public constructor
    }

    public static OptionsFragment newInstance() {
        OptionsFragment fragment = new OptionsFragment();
        Bundle args = new Bundle();

        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        mFirebaseHelper = new FirebaseHelper(getActivity());
        // Inflate the layout for this fragment
        View rootView = inflater.inflate(R.layout.fragment_options, container, false);

        autoPlaySwitch(rootView);
        autoBrightnessSwitch(rootView);
        powerConnectedSwitch(rootView);
        sendToBackgroundSwitch(rootView);
        waitTillOffPhoneSwitch(rootView);
        showNotificationSwitch(rootView);
        wifiOffUseTimeSpansSwitch(rootView);
        brightBrightnessButton(rootView);
        dimBrightnessButton(rootView);
        usePriorityModeSwitch(rootView);

        setFonts(rootView, getContext());
        setButtonPreferences(rootView, getActivity());
        setMaxVolumeSeekBar(rootView);
        setupRestoreOriginalVolumeCheckBox(rootView);

        wifiOffDeviceButton(rootView);

        setWifiUseTimeSpansCaption(rootView);

        return rootView;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
    }

    @Override
    public void onDetach() {
        super.onDetach();
    }

    @Override
    public void onResume(){
        super.onResume();
        if(mFirebaseHelper == null){
            mFirebaseHelper = new FirebaseHelper(getActivity());
        }
    }

    private void setWifiUseTimeSpansCaption(View view){
        String captionText;
        TextView wifiUseTimeSpanCaption = (TextView)view.findViewById(R.id.wifi_use_time_spans_desc);
        if(BAPMPreferences.INSTANCE.getCanLaunchDirections(getContext())){
            captionText = view.getResources().getString(R.string.wifi_use_time_spans_caption_work_home);
        } else {
            captionText = view.getResources().getString(R.string.wifi_use_time_spans_caption_morning_evening);
        }

        wifiUseTimeSpanCaption.setText(captionText);
    }

    private void setButtonPreferences(View view, Context context) {
        Boolean btnState;
        Switch setting_switch;

        btnState = BAPMPreferences.INSTANCE.getAutoPlayMusic(context);
        setting_switch = (Switch) view.findViewById(R.id.auto_play);
        setting_switch.setChecked(btnState);

        btnState = BAPMPreferences.INSTANCE.getPowerConnected(context);
        setting_switch = (Switch) view.findViewById(R.id.power_connected);
        setting_switch.setChecked(btnState);

        btnState = BAPMPreferences.INSTANCE.getSendToBackground(context);
        setting_switch = (Switch) view.findViewById(R.id.send_to_background);
        setting_switch.setChecked(btnState);

        btnState = BAPMPreferences.INSTANCE.getWaitTillOffPhone(context);
        setting_switch = (Switch) view.findViewById(R.id.wait_till_off_phone);
        setting_switch.setChecked(btnState);

        btnState = BAPMPreferences.INSTANCE.getShowNotification(context);
        setting_switch = (Switch) view.findViewById(R.id.show_notification);
        setting_switch.setChecked(btnState);

        btnState = BAPMPreferences.INSTANCE.getWifiUseMapTimeSpans(context);
        setting_switch = (Switch) view.findViewById(R.id.wifi_use_time_spans);
        setting_switch.setChecked(btnState);

        if(!PermissionHelper.isPermissionGranted(context, PermissionHelper.Permission.COARSE_LOCATION))
            BAPMPreferences.INSTANCE.setAutoBrightness(context, false);
        btnState = BAPMPreferences.INSTANCE.getAutoBrightness(context);
        setting_switch = (Switch) view.findViewById(R.id.auto_brightness);
        setting_switch.setChecked(btnState);

        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            btnState = BAPMPreferences.INSTANCE.getUsePriorityMode(context);
            setting_switch = (Switch) view.findViewById(R.id.sw_priority_mode);
            TextView switch_explaination = (TextView) view.findViewById(R.id.tv_priority_mode_explaination);

            setting_switch.setVisibility(View.VISIBLE);
            setting_switch.setChecked(btnState);

            switch_explaination.setVisibility(View.VISIBLE);
        }

    }

    public void wifiOffUseTimeSpansSwitch(View view){
        Switch autoPlaySwitch = (Switch)view.findViewById(R.id.wifi_use_time_spans);
        autoPlaySwitch.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                boolean on = ((Switch) view).isChecked();
                mFirebaseHelper.featureEnabled(FirebaseHelper.Option.WIFI_OFF_USE_TIME_SPANS, on);
                if (on) {
                    BAPMPreferences.INSTANCE.setWifiUseMapTimeSpans(getActivity(), true);
                    Log.d(TAG, "WIFI OFF Use Time Spans Switch is ON");
                } else {
                    BAPMPreferences.INSTANCE.setWifiUseMapTimeSpans(getActivity(), false);
                    Log.d(TAG, "WIFI OFF Use Time Spans Switch is OFF");
                }
            }
        });
    }

    public void autoPlaySwitch(View view){
        Switch autoPlaySwitch = (Switch)view.findViewById(R.id.auto_play);
        autoPlaySwitch.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                boolean on = ((Switch) view).isChecked();
                mFirebaseHelper.featureEnabled(FirebaseHelper.Option.PLAY_MUSIC, on);
                if (on) {
                    BAPMPreferences.INSTANCE.setAutoplayMusic(getActivity(), true);
                    Log.d(TAG, "AutoPlaySwitch is ON");
                } else {
                    BAPMPreferences.INSTANCE.setAutoplayMusic(getActivity(), false);
                    Log.d(TAG, "AutoPlaySwitch is OFF");
                }
            }
        });
    }

    public void powerConnectedSwitch(View view){
        Switch powerConnectedSwitch = (Switch)view.findViewById(R.id.power_connected);
        powerConnectedSwitch.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                boolean on = ((Switch) view).isChecked();
                mFirebaseHelper.featureEnabled(FirebaseHelper.Option.POWER_REQUIRED, on);
                if(on){
                    BAPMPreferences.INSTANCE.setPowerConnected(getActivity(), true);
                    Log.d(TAG, "PowerConnected Switch is ON");
                }else{
                    BAPMPreferences.INSTANCE.setPowerConnected(getActivity(), false);
                    Log.d(TAG, "PowerConnected Switch is OFF");
                }
            }
        });
    }

    public void sendToBackgroundSwitch(View view){
        Switch sendToBackgroundSwitch = (Switch)view.findViewById(R.id.send_to_background);
        sendToBackgroundSwitch.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                boolean on = ((Switch) view).isChecked();
                mFirebaseHelper.featureEnabled(FirebaseHelper.Option.GO_HOME, on);
                if(on){
                    BAPMPreferences.INSTANCE.setSendToBackground(getActivity(), true);
                    Log.d(TAG, "SendToBackground Switch is ON");
                }else{
                    BAPMPreferences.INSTANCE.setSendToBackground(getActivity(), false);
                    Log.d(TAG, "SendToBackground Switch is OFF");
                }
            }
        });
    }

    public void waitTillOffPhoneSwitch(View view){
        Switch waitTillOffPhoneSwitch = (Switch)view.findViewById(R.id.wait_till_off_phone);
        waitTillOffPhoneSwitch.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                boolean on = ((Switch) view).isChecked();
                mFirebaseHelper.featureEnabled(FirebaseHelper.Option.CALL_COMPLETED, on);
                if(on){
                    BAPMPreferences.INSTANCE.setWaitTillOffPhone(getActivity(), true);
                    Log.d(TAG, "WaitTillOffPhone Switch is ON");
                }else{
                    BAPMPreferences.INSTANCE.setWaitTillOffPhone(getActivity(), false);
                    Log.d(TAG, "WaitTillOffPhone Switch is OFF");
                }
            }
        });
    }

    public void showNotificationSwitch(View view) {
        Switch autoBrightnessSwitch = (Switch)view.findViewById(R.id.show_notification);
        autoBrightnessSwitch.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                boolean on = ((Switch) view).isChecked();
                mFirebaseHelper.featureEnabled(FirebaseHelper.Option.SHOW_NOTIFICATION, on);
                if(on){
                    BAPMPreferences.INSTANCE.setShowNotification(getActivity(), true);
                    Log.d(TAG, "Show Notification Switch is ON");
                }else{
                    BAPMPreferences.INSTANCE.setShowNotification(getActivity(), false);
                    Log.d(TAG, "Show Notification Switch is OFF");
                }
            }
        });
    }

    public void autoBrightnessSwitch(View view){
        Switch autoBrightnessSwitch = (Switch)view.findViewById(R.id.auto_brightness);
        autoBrightnessSwitch.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                boolean on = ((Switch) view).isChecked();
                mFirebaseHelper.featureEnabled(FirebaseHelper.Option.AUTO_BRIGHTNESS, on);
                if(on){
                    PermissionHelper.checkPermission(getActivity(), PermissionHelper.Permission.COARSE_LOCATION);

                    BAPMPreferences.INSTANCE.setAutoBrightness(getActivity(), true);

                    Log.d(TAG, "AutoBrightness Switch is ON");
                }else{
                    BAPMPreferences.INSTANCE.setAutoBrightness(getActivity(), false);
                    Log.d(TAG, "AutoBrightness Switch is OFF");
                }
            }
        });
    }

    public void usePriorityModeSwitch(View view) {
        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            Switch priorityModeSwitch = (Switch)view.findViewById(R.id.sw_priority_mode);
            priorityModeSwitch.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    boolean on = ((Switch) view).isChecked();
                    mFirebaseHelper.featureEnabled(FirebaseHelper.Option.PRIORITY_MODE, on);
                    if(on){
                        String permission = Manifest.permission.ACCESS_NOTIFICATION_POLICY;
                        PermissionHelper.checkPermission(getActivity(), permission);

                        BAPMPreferences.INSTANCE.setUsePriorityMode(getActivity(), true);

                        Log.d(TAG, "AutoBrightness Switch is ON");
                    }else{
                        BAPMPreferences.INSTANCE.setUsePriorityMode(getActivity(), false);
                        Log.d(TAG, "AutoBrightness Switch is OFF");
                    }
                }
            });
        }
    }

    public void dimBrightnessButton(View view){
        Button dimBrightnessButton = (Button)view.findViewById(R.id.dimtimebutton);
        dimBrightnessButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                DialogFragment newFragment = TimePickerFragment.newInstance(TimePickerFragment.TypeOfTimeSet.SCREEN_BRIGHTNESS_TIME, true, BAPMPreferences.INSTANCE.getDimTime(getActivity()), "Set Dim Time");
                newFragment.show(getActivity().getSupportFragmentManager(), "timePicker");
            }
        });
    }

    public void brightBrightnessButton(View view){
        Button brightBrightnessButton = (Button)view.findViewById(R.id.brighttimebutton);
        brightBrightnessButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                DialogFragment newFragment = TimePickerFragment.newInstance(TimePickerFragment.TypeOfTimeSet.SCREEN_BRIGHTNESS_TIME, false, BAPMPreferences.INSTANCE.getBrightTime(getActivity()), "Set Bright Time");
                newFragment.show(getActivity().getSupportFragmentManager(), "timePicker");
            }
        });
    }

    public void setMaxVolumeSeekBar(View view){
        AudioManager audioManager = (AudioManager)getActivity().getSystemService(Context.AUDIO_SERVICE);
        SeekBar volumeSeekBar = (SeekBar)view.findViewById(R.id.max_volume_seekBar);

        volumeSeekBar.setMax(audioManager.getStreamMaxVolume(AudioManager.STREAM_MUSIC));
        volumeSeekBar.setProgress(BAPMPreferences.INSTANCE.getUserSetMaxVolume(getActivity()));
        volumeSeekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                BAPMPreferences.INSTANCE.setUserSetMaxVolume(getActivity().getApplicationContext(), progress);
                Log.d(TAG, "User set MAX volume: " + Integer.toString(BAPMPreferences.INSTANCE.getUserSetMaxVolume(getActivity().getApplicationContext())));
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {

            }
        });
    }

    public void setupRestoreOriginalVolumeCheckBox(View view){
        final Context context = view.getContext();
        boolean isEnabled = BAPMPreferences.INSTANCE.getRestoreNotificationVolume(context);
        CheckBox restoreVolumeCheckBox = (CheckBox)view.findViewById(R.id.cb_restore_original_volume);

        restoreVolumeCheckBox.setChecked(isEnabled);

        restoreVolumeCheckBox.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                BAPMPreferences.INSTANCE.setRestoreNotificationVolume(context, isChecked);
            }
        });
    }

    public void wifiOffDeviceButton(View view){
        Button wifiOffDeviceButton = (Button)view.findViewById(R.id.wifi_off_button);
        wifiOffDeviceButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                mFirebaseHelper.selectionMade(FirebaseHelper.Selection.SET_WIFI_OFF_DEVICE);
                DialogFragment newFragment = WifiOffFragment.newInstance();
                newFragment.show(getActivity().getSupportFragmentManager(), "wifiOffFragment");
            }
        });
    }

    private void setFonts(View view, Context context){
        Typeface typeface_bold = Typeface.createFromAsset(context.getAssets(), "fonts/TitilliumText600wt.otf");

        TextView textView = (TextView)view.findViewById(R.id.settingsText);
        textView.setTypeface(typeface_bold);

        textView = (TextView)view.findViewById(R.id.auto_play);
        textView.setTypeface(typeface_bold);

        textView = (TextView)view.findViewById(R.id.power_connected);
        textView.setTypeface(typeface_bold);

        textView = (TextView)view.findViewById(R.id.send_to_background);
        textView.setTypeface(typeface_bold);

        textView = (TextView)view.findViewById(R.id.wait_till_off_phone);
        textView.setTypeface(typeface_bold);

        textView = (TextView)view.findViewById(R.id.auto_brightness);
        textView.setTypeface(typeface_bold);

        textView = (TextView)view.findViewById(R.id.manualBrtLabel);
        textView.setTypeface(typeface_bold);

        textView = (TextView)view.findViewById(R.id.userSetMaxVolumeLabel);
        textView.setTypeface(typeface_bold);

        textView = (TextView)view.findViewById(R.id.show_notification);
        textView.setTypeface(typeface_bold);

        textView = (TextView)view.findViewById(R.id.wifi_use_time_spans);
        textView.setTypeface(typeface_bold);
    }
}

File: app/src/main/java/maderski/bluetoothautoplaymusic/ui/fragments/OptionsFragment.kt
package maderski.bluetoothautoplaymusic.ui.fragments

import android.Manifest
import android.content.Context
import android.graphics.Typeface
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import maderski.bluetoothautoplaymusic.R
import maderski.bluetoothautoplaymusic.analytics.FirebaseHelper
import maderski.bluetoothautoplaymusic.analytics.constants.OptionConstants
import maderski.bluetoothautoplaymusic.controls.VolumeControl
import maderski.bluetoothautoplaymusic.permission.PermissionManager
import maderski.bluetoothautoplaymusic.sharedprefs.BAPMPreferences
import org.koin.android.ext.android.inject

class OptionsFragment : androidx.fragment.app.Fragment() {
    private val preferences: BAPMPreferences by inject()
    private val volumeControl: VolumeControl by inject()
    private val permissionManager: PermissionManager by inject()

    private lateinit var mFirebaseHelper: FirebaseHelper

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mFirebaseHelper = FirebaseHelper(requireActivity())
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val fragmentActivity = requireActivity()

        // Inflate the layout for this fragment
        val rootView = inflater.inflate(R.layout.fragment_options, container, false)

        autoPlaySwitch(rootView)
        autoBrightnessSwitch(rootView)
        powerConnectedSwitch(rootView)
        sendToBackgroundSwitch(rootView)
        waitTillOffPhoneSwitch(rootView)
        brightBrightnessButton(rootView)
        dimBrightnessButton(rootView)
        usePriorityModeSwitch(rootView)

        setFonts(rootView, fragmentActivity)
        setButtonPreferences(rootView, fragmentActivity)
        setMaxVolumeSeekBar(rootView)
        setupRestoreOriginalVolumeCheckBox(rootView)

        return rootView
    }

    private fun setButtonPreferences(view: View, context: Context) {
        var btnState: Boolean?
        var settingSwitch: Switch

        btnState = preferences.getAutoPlayMusic()
        settingSwitch = view.findViewById(R.id.auto_play)
        settingSwitch.isChecked = btnState

        btnState = preferences.getPowerConnected()
        settingSwitch = view.findViewById(R.id.power_connected)
        settingSwitch.isChecked = btnState

        btnState = preferences.getSendToBackground()
        settingSwitch = view.findViewById(R.id.send_to_background)
        settingSwitch.isChecked = btnState

        btnState = preferences.getWaitTillOffPhone()
        settingSwitch = view.findViewById(R.id.wait_till_off_phone)
        settingSwitch.isChecked = btnState

        if (!permissionManager.isLocationPermissionGranted())
            preferences.setAutoBrightness(false)
        btnState = preferences.getAutoBrightness()
        settingSwitch = view.findViewById(R.id.auto_brightness)
        settingSwitch.isChecked = btnState

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            btnState = preferences.getUsePriorityMode()
            settingSwitch = view.findViewById(R.id.sw_priority_mode)
            val switchExplaination = view.findViewById<TextView>(R.id.tv_priority_mode_explaination)

            settingSwitch.visibility = View.VISIBLE
            settingSwitch.isChecked = btnState

            switchExplaination.visibility = View.VISIBLE
        }

    }

    fun autoPlaySwitch(view: View) {
        val autoPlaySwitch = view.findViewById<Switch>(R.id.auto_play)
        autoPlaySwitch.setOnClickListener { autoPlaySwitchView ->
            val on = (autoPlaySwitchView as Switch).isChecked
            mFirebaseHelper.featureEnabled(OptionConstants.PLAY_MUSIC, on)
            if (on) {
                preferences.setAutoplayMusic(true)
                Log.d(TAG, "AutoPlaySwitch is ON")
            } else {
                preferences.setAutoplayMusic(false)
                Log.d(TAG, "AutoPlaySwitch is OFF")
            }
        }
    }

    fun powerConnectedSwitch(view: View) {
        val powerConnectedSwitch = view.findViewById<Switch>(R.id.power_connected)
        powerConnectedSwitch.setOnClickListener { powerConnectedSwitchView ->
            val on = (powerConnectedSwitchView as Switch).isChecked
            mFirebaseHelper.featureEnabled(OptionConstants.POWER_REQUIRED, on)
            if (on) {
                preferences.setPowerConnected(true)
                Log.d(TAG, "PowerConnected Switch is ON")
            } else {
                preferences.setPowerConnected(false)
                Log.d(TAG, "PowerConnected Switch is OFF")
            }
        }
    }

    fun sendToBackgroundSwitch(view: View) {
        val sendToBackgroundSwitch = view.findViewById<View>(R.id.send_to_background)
        sendToBackgroundSwitch.setOnClickListener { sendToBackgroundSwitchView ->
            val on = (sendToBackgroundSwitchView as Switch).isChecked
            mFirebaseHelper.featureEnabled(OptionConstants.GO_HOME, on)
            if (on) {
                preferences.setSendToBackground(true)
                Log.d(TAG, "SendToBackground Switch is ON")
            } else {
                preferences.setSendToBackground(false)
                Log.d(TAG, "SendToBackground Switch is OFF")
            }
        }
    }

    fun waitTillOffPhoneSwitch(view: View) {
        val waitTillOffPhoneSwitch = view.findViewById<View>(R.id.wait_till_off_phone)
        waitTillOffPhoneSwitch.setOnClickListener { waitTillOffPhoneSwitchView ->
            val on = (waitTillOffPhoneSwitchView as Switch).isChecked
            mFirebaseHelper.featureEnabled(OptionConstants.CALL_COMPLETED, on)
            if (on) {
                preferences.setWaitTillOffPhone(true)
                Log.d(TAG, "WaitTillOffPhone Switch is ON")
            } else {
                preferences.setWaitTillOffPhone(false)
                Log.d(TAG, "WaitTillOffPhone Switch is OFF")
            }
        }
    }

    fun autoBrightnessSwitch(view: View) {
        val autoBrightnessSwitch = view.findViewById<View>(R.id.auto_brightness)
        autoBrightnessSwitch.setOnClickListener { autoBrightnessSwitchView ->
            val on = (autoBrightnessSwitchView as Switch).isChecked
            mFirebaseHelper.featureEnabled(OptionConstants.AUTO_BRIGHTNESS, on)
            if (on) {
                permissionManager.checkLocationPermission(requireActivity())

                preferences.setAutoBrightness(true)

                Log.d(TAG, "AutoBrightness Switch is ON")
            } else {
                preferences.setAutoBrightness(false)
                Log.d(TAG, "AutoBrightness Switch is OFF")
            }
        }
    }

    fun usePriorityModeSwitch(view: View) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            val priorityModeSwitch = view.findViewById<View>(R.id.sw_priority_mode)
            priorityModeSwitch.setOnClickListener { priorityModeSwitchView ->
                val on = (priorityModeSwitchView as Switch).isChecked
                mFirebaseHelper.featureEnabled(OptionConstants.PRIORITY_MODE, on)
                if (on) {
                    permissionManager.checkAccessNotificationPolicyPermission(requireActivity())

                    preferences.setUsePriorityMode(true)

                    Log.d(TAG, "AutoBrightness Switch is ON")
                } else {
                    preferences.setUsePriorityMode(false)
                    Log.d(TAG, "AutoBrightness Switch is OFF")
                }
            }
        }
    }

    fun dimBrightnessButton(view: View) {
        val dimBrightnessButton = view.findViewById<Button>(R.id.dimtimebutton)
        dimBrightnessButton.setOnClickListener {
            val newFragment = TimePickerFragment.newInstance(TimePickerFragment.SCREEN_BRIGHTNESS_TIME, true, preferences.getDimTime(), "Set Dim Time")
            newFragment.show(requireActivity().supportFragmentManager, "timePicker")
        }
    }

    fun brightBrightnessButton(view: View) {
        val brightBrightnessButton = view.findViewById<Button>(R.id.brighttimebutton)
        brightBrightnessButton.setOnClickListener {
            val newFragment = TimePickerFragment.newInstance(TimePickerFragment.SCREEN_BRIGHTNESS_TIME, false, preferences.getBrightTime(), "Set Bright Time")
            newFragment.show(requireActivity().supportFragmentManager, "timePicker")
        }
    }

    fun setMaxVolumeSeekBar(view: View) {
        val deviceMaxVolume = volumeControl.getDeviceMaxVolume()
        val volumeSeekBar = view.findViewById<SeekBar>(R.id.max_volume_seekBar)

        volumeSeekBar.max = volumeControl.getDeviceMaxVolume()
        volumeSeekBar.progress = preferences.getUserSetMaxVolume(deviceMaxVolume)
        volumeSeekBar.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                preferences.setUserSetMaxVolume(progress)
                Log.d(TAG, "User set MAX volume: " + Integer.toString(preferences.getUserSetMaxVolume(deviceMaxVolume)))
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {

            }

            override fun onStopTrackingTouch(seekBar: SeekBar) {

            }
        })
    }

    fun setupRestoreOriginalVolumeCheckBox(view: View) {
        val isEnabled = preferences.getRestoreNotificationVolume()
        val restoreVolumeCheckBox = view.findViewById<CheckBox>(R.id.cb_restore_original_volume)

        restoreVolumeCheckBox.isChecked = isEnabled

        restoreVolumeCheckBox.setOnCheckedChangeListener { buttonView, isChecked -> preferences.setRestoreNotificationVolume(isChecked) }
    }

    private fun setFonts(view: View, context: Context) {
        val typeface_bold = Typeface.createFromAsset(context.assets, "fonts/TitilliumText600wt.otf")

        var textView = view.findViewById<TextView>(R.id.settingsText)
        textView.typeface = typeface_bold

        textView = view.findViewById(R.id.auto_play)
        textView.typeface = typeface_bold

        textView = view.findViewById(R.id.power_connected)
        textView.typeface = typeface_bold

        textView = view.findViewById(R.id.send_to_background)
        textView.typeface = typeface_bold

        textView = view.findViewById(R.id.wait_till_off_phone)
        textView.typeface = typeface_bold

        textView = view.findViewById(R.id.auto_brightness)
        textView.typeface = typeface_bold

        textView = view.findViewById(R.id.manualBrtLabel)
        textView.typeface = typeface_bold

        textView = view.findViewById(R.id.userSetMaxVolumeLabel)
        textView.typeface = typeface_bold
    }

    companion object {
        private val TAG = "OptionsFragment"

        fun newInstance(): OptionsFragment {
            val fragment = OptionsFragment()
            val args = Bundle()

            fragment.arguments = args
            return fragment
        }
    }
}// Required empty public constructor
